src/tools/rbd_cache/CacheControllerSocketClient.hpp

```diff
@@ -19,32 +22,64 @@ namespace cache {
 
 class CacheClient {
 public:
-  CacheClient(boost::asio::io_service& io_service,
-              const std::string& file, ClientProcessMsg processmsg)
-    : io_service_(io_service),
-      io_service_work_(io_service),
-      socket_(io_service),
+  CacheClient(const std::string& file, ClientProcessMsg processmsg, CephContext* ceph_ctx)
+    : m_io_service_work(m_io_service),
+      m_dm_socket(m_io_service),
       m_client_process_msg(processmsg),
-      ep_(stream_protocol::endpoint(file))
+      m_ep(stream_protocol::endpoint(file)),
+      m_session_work(false),
+      cct(ceph_ctx)
   {
-     io_thread.reset(new std::thread([this](){io_service_.run(); }));
+     // TODO wrapper io_service
+     std::thread thd([this](){
+                      m_io_service.run();});
+     thd.detach();
```

Comment:
Why the `thread.detach()` approach is adopted?  I'm new to `std::thread` and `boost::asio`, but here the worker thread is not totally independent to its parent. The `m_io_service` object might be destroyed while the worker thread is running, that might be the reason of the following call stack:

The `scheduler::task_operation_` is resided in  `scheduler::op_queue_` while the scheduler's destructor is called.
```
#0  0x0000000000000000 in ?? ()
#1  0x00007f61917710d9 in boost::asio::detail::scheduler_operation::destroy (this=0x7f614c0026a0)
    at /mnt/test/ceph-readcache/build/boost/include/boost/asio/detail/scheduler_operation.hpp:45
#2  boost::asio::detail::op_queue_access::destroy<boost::asio::detail::scheduler_operation> (o=0x7f614c0026a0)
    at /mnt/test/ceph-readcache/build/boost/include/boost/asio/detail/op_queue.hpp:47
#3  boost::asio::detail::op_queue<boost::asio::detail::scheduler_operation>::~op_queue (this=0x7f614c0026c8, __in_chrg=<optimized out>)
    at /mnt/test/ceph-readcache/build/boost/include/boost/asio/detail/op_queue.hpp:81
#4  boost::asio::detail::scheduler::~scheduler (this=0x7f614c0025f0, __in_chrg=<optimized out>)
    at /mnt/test/ceph-readcache/build/boost/include/boost/asio/detail/scheduler.hpp:38
#5  boost::asio::detail::scheduler::~scheduler (this=0x7f614c0025f0, __in_chrg=<optimized out>)
    at /mnt/test/ceph-readcache/build/boost/include/boost/asio/detail/scheduler.hpp:38
#6  0x00007f61916ca2da in boost::asio::detail::service_registry::destroy (service=<optimized out>)
    at /mnt/test/ceph-readcache/build/boost/include/boost/asio/detail/impl/service_registry.ipp:110
#7  boost::asio::detail::service_registry::destroy_services (this=0x7f614c0025a0)
    at /mnt/test/ceph-readcache/build/boost/include/boost/asio/detail/impl/service_registry.ipp:54
#8  boost::asio::execution_context::destroy (this=0x7f614c0020b0)
    at /mnt/test/ceph-readcache/build/boost/include/boost/asio/impl/execution_context.ipp:46
#9  boost::asio::execution_context::~execution_context (this=0x7f614c0020b0, __in_chrg=<optimized out>)
    at /mnt/test/ceph-readcache/build/boost/include/boost/asio/impl/execution_context.ipp:35
#10 boost::asio::io_context::~io_context (this=0x7f614c0020b0, __in_chrg=<optimized out>)
    at /mnt/test/ceph-readcache/build/boost/include/boost/asio/impl/io_context.ipp:55
#11 ceph::immutable_obj_cache::CacheClient::~CacheClient (this=0x7f614c0020b0, __in_chrg=<optimized out>)
    at /mnt/test/ceph-readcache/src/tools/ceph_immutable_object_cache/CacheClient.h:23
#12 librbd::cache::SharedReadOnlyObjectDispatch<librbd::ImageCtx>::~SharedReadOnlyObjectDispatch (this=<optimized out>, __in_chrg=<optimized out>)
    at /mnt/test/ceph-readcache/src/librbd/cache/SharedReadOnlyObjectDispatch.cc:34
#13 0x00007f61916ca7f9 in librbd::cache::SharedReadOnlyObjectDispatch<librbd::ImageCtx>::~SharedReadOnlyObjectDispatch (this=0x7f614c001a60, 
    __in_chrg=<optimized out>) at /mnt/test/ceph-readcache/src/librbd/cache/SharedReadOnlyObjectDispatch.cc:35
#14 0x00007f61916ff1d9 in librbd::io::ObjectDispatcher<librbd::ImageCtx>::shut_down_object_dispatch(librbd::io::ObjectDispatcher<librbd::ImageCtx>::ObjectDispatchMeta&, Context**)::{lambda(int)#1}::operator()(int) const (r=0, __closure=0x7f614c004720)
    at /mnt/test/ceph-readcache/src/librbd/io/ObjectDispatcher.cc:254
#15 boost::detail::function::void_function_obj_invoker1<librbd::io::ObjectDispatcher<librbd::ImageCtx>::shut_down_object_dispatch(librbd::io::ObjectDispatcher<librbd::ImageCtx>::ObjectDispatchMeta&, Context**)::{lambda(int)#1}, void, int>::invoke(boost::detail::function::function_buffer&, int) (
    function_obj_ptr=..., a0=0) at /mnt/test/ceph-readcache/build/boost/include/boost/function/function_template.hpp:159
#16 0x00007f6191632e0c in boost::function1<void, int>::operator() (a0=<optimized out>, this=<optimized out>)
    at /mnt/test/ceph-readcache/build/boost/include/boost/function/function_template.hpp:768
#17 FunctionContext::finish (this=<optimized out>, r=<optimized out>) at /mnt/test/ceph-readcache/src/include/Context.h:489
#18 0x00007f619161a629 in Context::complete (this=0x7f614c004710, r=<optimized out>) at /mnt/test/ceph-readcache/src/include/Context.h:76
#19 0x00007f6187534dcb in ThreadPool::worker (this=0xbe1da0, wt=<optimized out>) at /mnt/test/ceph-readcache/src/common/WorkQueue.cc:120
#20 0x00007f6187535800 in ThreadPool::WorkThread::entry (this=<optimized out>) at /mnt/test/ceph-readcache/src/common/WorkQueue.h:449
#21 0x00007f61904ce6ba in start_thread (arg=0x7f615f7fe700) at pthread_create.c:333
#22 0x00007f619000041d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
```

Reply:
Why the thread.detach() approach is adopted?
  --> later, we will wrapper io_service to to implement callback threadpool. Current implements is rough and simple. This is why we write TODO here. 

 the worker thread is not totally independent to its parent. 
   --> Due to boost::asio::io_service::work(m_io_service),  although scheduler::op_queue_ is empty,  io_service still wait for new scheduler::task_operaion_, namely m_io_service.run() will never end,  namely the lambda function of std::thread's body never actively end. So, although thd is local variable, but it still can works. 

current implement is similar with the following demo: 

 
   #include <iostream>
   #include <thread>
   #include <unistd.h>
   #include <boost/asio.hpp>
  
  using namespace std;
  
   int main() {
 
      boost::asio::io_service m_io_service;
     boost::asio::io_service::work worker(m_io_service);
 
      {
         std::thread thd([&m_io_service](){
             std::cout<<"thread start..."<<std::endl;
              m_io_service.run();
              std::cout<<"thread over..."<<std::endl;
          });
         thd.detach();
      }
 
      while(1) {
          cout << "sleep ...." << endl;
          sleep(10);
      }
 
 
     return 0;
  }

Reply:
> later, we will wrapper io_service to to implement callback threadpool. Current implements is rough and simple. This is why we write TODO here.

I see. But for the current implementation, should we replace `thread::detach()` by `io_service::stop()` and `thread::join()`? (see the `~CacheClient()` in commit 9f8ff821).
Otherwise, the worker thread cannot terminate properly (e.g., "thread over..." might not be executed).

I also have question about the cache eviction method. Why the eviction is done by a background thread, instead of done by IO threads? Also, is it good to keep the cache usage under a threshold?

Finally, is there any further planning or schedule on this project? I'm interested in this project. Is there any task item that I could participant in?

Reply:
yes, your method is more batter choice. thanks for your suggestions. 
For this callback thread, i need to talk with zhouyuan. our init plan is to implement an callback-threadpool. there have one refer: https://github.com/baidu/sofa-pbrpc/blob/master/src/sofa/pbrpc/thread_group_impl.h 

based on this refers, we can control callback thread num, concurrent, thread::join, and post function. 
Maybe next month, i will have time to do this feature. 
----
For evict thread, we want to split promoting and evicting. 

--
For plan, please wait for zhouyuan's response. 

In the near future, we will focus on stability, performance, and fix some bug. 
we still have some pending code to need merge this branch.

Reply:
> For evict thread, we want to split promoting and evicting.

Can you be more specific about the motivation behind splitting promotion and eviction? I don't really understand the necessity in splitting the two mechanisms. IMO instead of having a seperate thread constantly checking if the cache is full, it's probably more straightforward to do eviction when a full cache was detected during promotion.

Reply:
> For evict thread, we want to split promoting and evicting.

Can you be more specific about the motivation behind splitting promotion and eviction? I don't really understand the necessity in splitting the two mechanisms. IMO instead of having a seperate thread constantly checking if the cache is full, it's probably more straightforward to do eviction when a full cache was detected during promotion.

Reply:
@usefulalgorithm thanks for the idea, agree that's more simpler and easier. 
The initial concern was for the performance - note we have only single thread for cache controller now. If multiple VMs are started on the same node, the single threaded promotion + eviction maybe some bottleneck. So the original plan is to create a thread pool for cache controller, which will do promotion + eviction work.  Will see if this really fix the performance issue on the above case

Reply:
@mingnus looks like I missed that when rebasing, will fix, thanks!

Reply:
@zhouyuan How about using block-based cache? If cached objects were written to raw device directly, then cache eviction is merely a metadata update, which should be faster than the current file-deletion implementation.

Reply:
@mingnus right! The objectcachestore can be improved to use block cache(similar as filestore vs. bluestore). In the design stage we had some discussion on this( http://pad.ceph.com/p/rbd_persistent_cache ). Currently our goal is to land the code the into master so we could have the basic layout first. We plan to improve this feature in the next PR.